### ISSUE TITLE
[java] New rule for MessageDigest.getInstance

### ISSUE THREAD

**Issue by boris-petrov** [pmd/pmd#1862]
```java
static MessageDigest INSTANCE = MessageDigest.getInstance("SHA-512");

INSTANCE.digest(...);
```

This is not thread-safe code. That is, one should not cache the `MessageDigest.getInstance` instance and reuse it multiple times between threads. Similar to the [UnsynchronizedStaticFormatter](https://pmd.github.io/pmd-6.14.0/pmd_rules_java_multithreading.html#unsynchronizedstaticformatter) check.

I was bit by this and that's why I'm suggesting it as a new check - it might help someone. :)

**Comment by AnthonyKot** [comment]
https://github.com/pmd/pmd/pull/1963

**Comment by adangel** [comment]
Hi @boris-petrov ,
I have a question for this. Looking into the issue more closely and looking at the PR #1963, I'm wondering whether there is a real use case of sharing a MessageDigest instance.

How did you fix your code? Did you use synchronized? Or would it be possible to not share the instance and just create it in the method, where you want to calculate the hash?
The latter case would have the benefit, that you don't need to think about thread safety, because then the instance is only accessible by one thread.

> That is, one should not cache the MessageDigest.getInstance instance and reuse it multiple 
> times between threads.

I think, that is exactly , what we should aiming for: Flagging the cases, where one caches a MessageDigest instance.

**Comment by boris-petrov** [comment]
@adangel - sorry for the late response, I somehow missed that comment.

I fixed it by creating a new instance every time I need it. Not that the other way (using synchronized) is more difficult or something, I just guess that `MessageDigest.getInstance` is "fast enough" so this was the better way in our case.

**Comment by adangel** [comment]
@boris-petrov Thanks for your response. Yes, I hoped, that this is the solution: just creating a new instance of MessageDigest...
